### PR TITLE
morph: use Notary Actor for notary requests submission

### PR DESCRIPTION
**Pull request by AnnaShaleva**
This PR is a tiny part of neo-go version upgrade.

And adjust the way how newly-created SN notary request is defined. Notary actor doesn't fill dummy witness for multisignature account (and that's the way how the system is supposed to function, see the documentation 7-th instruction at
https://github.com/nspcc-dev/neo-go/blob/master/docs/notary.md#2-request-submission). Thus, to check the request was just received from the storage node we must ensure the inner ring witness has empty invocation script.

That was the reason of https://github.com/nspcc-dev/neo-go/issues/2955.

Refs. https://github.com/nspcc-dev/neo-go/pull/2981.

TODO:

- [x] Need to resolve TODOs marked in code, @roman-khimov, need your decision.
- [x] Update CHANGELOG.

### CODE REVIEW

**Comment by AnnaShaleva** [comment]
Checked on the dev-env against the https://github.com/nspcc-dev/neo-go/issues/2955, SNs are able to properly enter the netmap:
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neofs-node$ ./bin/neofs-cli netmap snapshot -g -r s01.neofs.devenv:8080
Epoch: 2
Node 1: 022bb4041c50d607ff871dec7e4cd7778388e0ea6849d84ccbd9aa8f32e16a8131 ONLINE /dns4/s01.neofs.devenv/tcp/8080 
	Continent: Europe
	Country: Russia
	CountryCode: RU
	Location: Moskva
	Price: 22
	SubDiv: Moskva
	SubDivCode: MOW
	UN-LOCODE: RU MOW
	User-Agent: NeoFS\/0.33
Node 2: 02ac920cd7df0b61b289072e6b946e2da4e1a31b9ab1c621bb475e30fa4ab102c3 ONLINE /dns4/s03.neofs.devenv/tcp/8080 
	Continent: Europe
	Country: Sweden
	CountryCode: SE
	Location: Stockholm
	Price: 11
	SubDiv: Stockholms l�n
	SubDivCode: AB
	UN-LOCODE: SE STO
	User-Agent: NeoFS\/0.33
Node 3: 038c862959e56b43e20f79187c4fe9e0bc7c8c66c1603e6cf0ec7f87ab6b08dc35 ONLINE /dns4/s04.neofs.devenv/tcp/8082/tls /dns4/s04.neofs.devenv/tcp/8080 
	Continent: Europe
	Country: Finland
	CountryCode: FI
	Location: Helsinki (Helsingfors)
	Price: 44
	SubDiv: Uusimaa
	SubDivCode: 18
	UN-LOCODE: FI HEL
	User-Agent: NeoFS\/0.33
Node 4: 03ff65b6ae79134a4dce9d0d39d3851e9bab4ee97abf86e81e1c5bbc50cd2826ae ONLINE /dns4/s02.neofs.devenv/tcp/8080 
	Continent: Europe
	Country: Russia
	CountryCode: RU
	Location: Saint Petersburg (ex Leningrad)
	Price: 33
	SubDiv: Sankt-Peterburg
	SubDivCode: SPE
	UN-LOCODE: RU LED
	User-Agent: NeoFS\/0.33
```

File: pkg/morph/client/notary.go [hunk 1]
```diff
@@ -411,26 +415,95 @@ func (c *Client) NotarySignAndInvokeTX(mainTx *transaction.Transaction) error {
...
-	)
+	// Here we need to add a committee signature (second witness) to the pre-validated
+	// main transaction without creating a new one and without parsing transaction's
+	// signers. However, Notary actor demands the proper set of signers for constructor,
+	// thus, fill it from the main transaction's signers list.
+	//
+	// TODO: @roman-khimov, this list of signers is used only for notary Actor constructor,
+	//  after that it's useless. Thus, I'd skip the signers parsing down below and create
+	//  some fake account instead to pass the Actor constructor constraint.
+	s := make([]actor.SignerAccount, 0, 3)
```
Comment (roman-khimov) [inline pkg/morph/client/notary.go:426]: `2, 3` `[0] = ... [1] = ...`

File: pkg/morph/client/notary.go [hunk 2]
```diff
@@ -411,26 +415,95 @@ func (c *Client) NotarySignAndInvokeTX(mainTx *transaction.Transaction) error {
...
+		{
+			// Proxy contract that will pay for the execution.
+			Signer:  mainTx.Signers[0],
+			Account: notary.FakeContractAccount(mainTx.Signers[0].Account),
+		},
+		{
+			// Inner ring multisignature.
+			Signer:  mainTx.Signers[1],
+			Account: multiaddrAccount,
+		},
```
Comment (roman-khimov) [inline pkg/morph/client/notary.go:437]: We have this set of signers used quite extensively, so I'd save an actor for them in some client context.

File: pkg/morph/client/notary.go [hunk 3]
```diff
@@ -411,26 +415,95 @@ func (c *Client) NotarySignAndInvokeTX(mainTx *transaction.Transaction) error {
...
+	}...)
+
+	if len(mainTx.Signers) > 3 {
+		// Invoker signature (simple signature account of storage node is expected).
```
Comment (roman-khimov) [inline pkg/morph/client/notary.go:441]: Then a fake one works fine as well, we're only supposed to be adding IR signatures now (we don't have any other cases for now), so we don't care who is the third signer, a correct signature is supposed to be present there anyway.

File: pkg/morph/client/notary.go [hunk 4]
```diff
@@ -411,26 +415,95 @@ func (c *Client) NotarySignAndInvokeTX(mainTx *transaction.Transaction) error {
...
+	}
+
+	// TODO: use c.notary.fallbackTime to set the fallback's NotValidBefore attribute.
+	//
+	// @roman-khimov, Currently Actor doesn't allow to customize fallback's NVB, see the
+	// https://github.com/nspcc-dev/neo-go/blob/7bc9b6e3236e4e89aea3ef86240746f3d17e591f/pkg/rpcclient/notary/actor.go#L278
+	// We may either adjust the Actor or create fallback manually right here.
+	// Honestly, I'd suggest to completely remove `fallbackTime` from the neofs-node's
+	// code/config in this PR, it's just not needed and not used anywhere except notary request creation.
+	// ACK removal or deprecation of `fallbackTime`?
```
Comment (roman-khimov) [inline pkg/morph/client/notary.go:496]: Yeah, it makes zero sense as a configuration option.

File: pkg/morph/client/notary.go [hunk 5]
```diff
@@ -496,199 +530,85 @@ func (c *Client) notaryInvoke(committee, invokedByAlpha bool, contract util.Uint
...
 	}
 
-	// add network fee for cosigners
-	//nolint:staticcheck // waits for neo-go v0.99.3 with notary actors
-	err = c.client.AddNetworkFee(
-		mainTx,
-		notaryFee,
-		c.notaryAccounts(invokedByAlpha, multiaddrAccount)...,
-	)
+	nAct, err := notary.NewActor(c.client, cosigners, c.acc)
```
Comment (roman-khimov) [inline pkg/morph/client/notary.go:538]: Again, this actor can and should be used long-term, not recreated for every request.

File: pkg/morph/client/notary.go [hunk 6]
```diff
@@ -496,199 +530,85 @@ func (c *Client) notaryInvoke(committee, invokedByAlpha bool, contract util.Uint
...
 	c.logger.Debug("notary request invoked",
 		zap.String("method", method),
-		zap.Uint32("valid_until_block", until),
-		zap.Uint32("fallback_valid_for", c.notary.fallbackTime),
-		zap.Stringer("tx_hash", resp.Hash().Reverse()))
+		zap.Uint32("valid_until_block", untilActual),
+		zap.String("tx_hash", mainH.StringLE()),
+		zap.String("fallback_hash", fbH.StringLE()))
 
 	return nil
```
Comment (roman-khimov) [inline pkg/morph/client/notary.go:564]: So we never track `mainH` and `fbH` further, right?

File: pkg/morph/client/notary.go [hunk 7]
```diff
@@ -496,199 +530,85 @@ func (c *Client) notaryInvoke(committee, invokedByAlpha bool, contract util.Uint
...
-	s = append(s, transaction.Signer{
-		Account:          hash.Hash160(multisigScript),
-		Scopes:           c.cfg.signer.Scopes,
-		AllowedContracts: c.cfg.signer.AllowedContracts,
-		AllowedGroups:    c.cfg.signer.AllowedGroups,
-	})
+		return nil, err
+	}
+	s := make([]actor.SignerAccount, 0, 3)
+	s = append(s, []actor.SignerAccount{
```
Comment (roman-khimov) [inline pkg/morph/client/notary.go:573]: It's the same as in another function.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/morph/client/notary.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/morph/client/notary.go:441]: The problem is that to create a proper `notary.FakeSimpleAccount(pub)` we need to parse existing verification script to get the `pub` out of it. The same problem with multisig account. And I'd like to completely avoid parsing, it's just not needed here and this signer is unesed anyway.

File: pkg/morph/client/notary.go (hunk 5 above)
Comment (AnnaShaleva) [reply pkg/morph/client/notary.go:538]: Then we have to track the committee updates somehow to recreate the actor on update.

File: pkg/morph/client/notary.go (hunk 6 above)
Comment (AnnaShaleva) [reply pkg/morph/client/notary.go:564]: Never. Just manually in logs during debugging.

File: pkg/morph/client/notary.go (hunk 7 above)
Comment (AnnaShaleva) [reply pkg/morph/client/notary.go:573]: It's not the same, and we can't reuse it. This one provides non-dummy sender account, and the other one provides non-dummy IR account (if needed). And we can't use `invokedByAlpha` flag for it to distinguish, it's responsible for other situation.

File: pkg/morph/client/notary.go (hunk 5 above)
Comment (roman-khimov) [reply pkg/morph/client/notary.go:538]: This should be done anyway. If it's not, well, we need an issue.

File: pkg/morph/client/notary.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/morph/client/notary.go:426]: Fixed.

File: pkg/morph/client/notary.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/morph/client/notary.go:437]: https://github.com/nspcc-dev/neofs-node/issues/2314 for that.

File: pkg/morph/client/notary.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/morph/client/notary.go:496]: Removed.

File: pkg/morph/client/notary.go (hunk 5 above)
Comment (AnnaShaleva) [reply pkg/morph/client/notary.go:538]: Added an issue, it's kind of a separate feature which requires some more attention to carefully watch the committee updates and keep Notary actor in sync with it.

File: pkg/morph/client/notary.go (hunk 6 above)
Comment (AnnaShaleva) [reply pkg/morph/client/notary.go:564]: BTW, I didn't change anything existing. I just added fallback hash to logs.

**Review by roman-khimov — APPROVED** [review]

File: pkg/morph/event/notary_preparator_test.go [hunk 8]
```diff
@@ -24,8 +24,8 @@ var (
 	alphaKeys      keys.PublicKeys
 	wrongAlphaKeys keys.PublicKeys
 
-	dummyInvocationScript      = append([]byte{byte(opcode.PUSHDATA1), 64}, make([]byte, 64)...)
-	wrongDummyInvocationScript = append([]byte{byte(opcode.PUSHDATA1), 64, 1}, make([]byte, 63)...)
+	dummyAlphabetInvocationScript = []byte{} // expected to be empty, as requester can't fill it in.
```
Comment (roman-khimov) [inline pkg/morph/event/notary_preparator_test.go:27]: We can use both new and old one in tests.

Comment (AnnaShaleva) [reply pkg/morph/event/notary_preparator_test.go]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**